File: problems/leetcode/0700/759/759.cpp
// 759. Employee Free Time
// https://leetcode.com/problems/employee-free-time/description/

using namespace std;

#include <queue>
#include <vector>

// Definition for an Interval.
class Interval {
public:
  int start;
  int end;

  Interval() {}

  Interval(int _start, int _end) {
    start = _start;
    end = _end;
  }
};

class Solution {
public:
  vector<Interval> employeeFreeTime(vector<vector<Interval>> schedule) {
    priority_queue<pair<int, int>, vector<pair<int, int>>,
                   greater<pair<int, int>>>
        pq;
    for (vector<Interval> &employee : schedule) {
      for (Interval &interval : employee) {
        pq.push({interval.start, interval.end});
      }
    }

    vector<pair<int, int>> overlaps;
    int start = pq.top().first, end = pq.top().second;
    pq.pop();
    while (!pq.empty()) {
      int nextStart = pq.top().first, nextEnd = pq.top().second;
      pq.pop();
      if (nextStart <= end) {
        end = max(end, nextEnd);
      } else {
        overlaps.push_back({start, end});
        start = nextStart;
        end = nextEnd;
      }
    }
    overlaps.push_back({start, end});

    vector<Interval> freeTimes;
    for (int i = 1; i < overlaps.size(); i++) {
      freeTimes.push_back(Interval(overlaps[i - 1].second, overlaps[i].first));
    }
    return freeTimes;
  }
};

// 1. 모든 schedule을 Priority Queue에 넣는다.
// 2. Priority Queue에서 하나씩 빼면서 합칠 수 있는 구간이라면 계속 합친다.
// 3. 합쳐진 schedule의 사이 구간들을 반환한다.

File: problems/leetcode/0700/759/759.py
# 759. Employee Free Time
# https://leetcode.com/problems/employee-free-time/description/

from collections import heapq
from typing import *

# 각 Employee별로 가장 일찍 시작하는 Interval을 계속 확인한다
# 마지막 Interval의 end보다 그 다음 Interval이 더 늦게 시작하면, common free time으로 간주 가능
# Priority Queue에 (시작 시간, employeeIndex, intervalIndex)를 저장
# 시작은 end = 0으로 간주 가능

# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end

class Solution:
    def employeeFreeTime(self, schedule: 'List[List[Interval]]') -> 'List[Interval]':
        pq = [(employee[0].start, i, 0) for i, employee in enumerate(schedule)] # heapq
        heapq.heapify(pq)

        freeTimes = []
        currentEnd = pq[0][0]
        while len(pq) > 0:
            _, employeeIndex, intervalIndex = heapq.heappop(pq)
            interval = schedule[employeeIndex][intervalIndex]
            if currentEnd < interval.start:
                freeTimes.append(Interval(currentEnd, interval.start))

            currentEnd = max(currentEnd, interval.end)
            if intervalIndex < len(schedule[employeeIndex]) - 1:
                heapq.heappush(pq, (schedule[employeeIndex][intervalIndex + 1].start, employeeIndex, intervalIndex + 1))

        return freeTimes
